### ISSUE TITLE
Plugin causes Npp crash with Npp v7.9.3

### ISSUE THREAD

**Issue by dinkumoil** [Hsilgos/nppsaveasadmin#24]
# Description

Notepad++ v7.9.3 crashes at startup when plugin _NppSaveAsAdmin_ is installed.

# Steps to Reproduce

1. Download and unpack a portabe version of Notepad++ v7.9.3
2. Install plugin _NppSaveAsAdmin_ using _PluginsAdmin_.
3. When Notepad++ attempts to restart, it crashes.

# Applies to

Notepad++ v7.9.3 (32 bit and 64 bit versions)
_NppSaveAsAdmin_ plugin v1.0.195.0

**Comment by dartraiden** [comment]
v 1.0.172 - ok
v 1.0.176 - crash

```
RAX : 000000000000BD40
RBX : 0000000000000000
RCX : 0000000000000013
RDX : 0000000000000008
RBP : 000000B95D263FD0     &"MZђ"
RSP : 000000B95D263E90
RSI : 000000B95D263EB0     "csmа\x01"
RDI : 000000B95D263850
R8  : 000000000000007F
R9  : 00000240915A2E70     "оюоюоюоюоюоюоюоюоюоюоюоюоюоюоюоюоюоюоюоюоюоюоюоюN"
R10 : 000000000000BCC0
R11 : 00000240915A35A0     "Injected function 'GetSaveFileNameW' in module 'Comdlg32.dll' was not found"
R12 : 0000000000000010
R13 : 000000B95D2641C0
R14 : 00007FFE01CAEE08     "Comdlg32.dll"
R15 : 00007FFE01CAEDE0     "GetSaveFileNameW"
RIP : 00007FFE102DD759     kernelbase.00007FFE102DD759
RFLAGS : 0000000000000204     L'Ȅ'
```

**Comment by allenatdecisiv** [comment]
Confirmed here as well with Save as admin 1.0.195 with Notepad++ 7.9.3.0 on Windows 10 20H2 Version 10.0.19042.804.

**Comment by mere-human** [comment]
The plugin tries to throw this exception from DllMain:
 "Injected function 'GetSaveFileNameW' in module 'Comdlg32.dll' was not found".
 But that leads to crash.

**Comment by mere-human** [comment]
Probably, Comdlg32.dll isn't in Notepad++ dependencies anymore.
Probably due to https://github.com/notepad-plus-plus/notepad-plus-plus/commit/e7079d57c6012bfc7cb67f6050bc8b563eb9758d

**Comment by mere-human** [comment]
This is the dependencies for Notepad++ v. 7.9.2:
![image](https://user-images.githubusercontent.com/9664141/107989039-073e2900-6fda-11eb-87b1-a6c1ae33957d.png)
And this is the dependencies for v. 7.9.3:
![image](https://user-images.githubusercontent.com/9664141/107989068-14f3ae80-6fda-11eb-9012-6f28451c4b54.png)
Note that GetSaveFileNameW, GetOpenFileNameW are not used anymore.

Tool used: https://github.com/lucasg/Dependencies

Also, throwing an exception from DllMain is a bad idea.

**Comment by pebl-hank** [comment]
@mere-human could you provide a temporary binary somewhere?
I'd really appreciate it.

**Comment by mere-human** [comment]
> @mere-human could you provide a temporary binary somewhere?
> I'd really appreciate it.

It would be nice to reach out the author @Hsilgos of the plugin somehow. So the PR can be merged.
If only he had left a public email...

If not, then I'll figure something out.
What is the best way to share the binaries?

**Comment by pebl-hank** [comment]
Couldn't you just include it into your fork?

**Comment by dinkumoil** [comment]
@mere-human 
Nice to hear that you took some effort to fix that issue. **BUT:** As long as you do not understand what exactly **causes** the exception to be thrown, this fix is not very reliable. Maybe the exception is thrown while retrieving some information needed for correct function of the plugin. If you only catch the exception, this information may be missing at a later point in time when the plugin tries to do its job **CORRECTLY**.

@mere-human , @pebl-hank and all
Since @dartraiden already figured out that [>>> version 1.0.172 <<<](https://github.com/Hsilgos/nppsaveasadmin/releases/tag/1.0.172) of the plugin still works with Notepad++ v7.9.3, I'd like to recommend to use this version instead.

**Comment by pebl-hank** [comment]
Works for me. Thanks for the suggestion!

**Comment by mere-human** [comment]
> As long as you do not understand what exactly **causes** the exception to be thrown, this fix is not very reliable.

I perfectly understand what causes the exception.
Also, I know for sure, that the DLL _should not throw from the DllMain() under any circumstances_.
This crashes the host application.

So in this case, if something bad had happened during the initialization we return false, the host app handles that, shows an error and continue functioning. All is good.

**Comment by dinkumoil** [comment]
@mere-human 

> So in this case, if something bad had happened during the initialization we return false, the host app handles that, shows an error and continue functioning.

Ah, OK. For sure, that's a legitimate solution and also reliable. It also fixes the flaw of not catching exceptions in _DllMain_. But users are still left with a non-functional plugin. So, downgrading to version 1.0.172 IMO is the best.

**Comment by allenatdecisiv** [comment]
@dinkumoil,

So those folks who are using the new (ish) Plugins Admin feature are to do what exactly? Manually install an old version and always remember that they can't update the "Save as admin" plugin?

How about reverting the changes to 1.0.195 and issue a new release so that end-users aren't presented with a confusing issue to resolve?

**Comment by dinkumoil** [comment]
@allenatdecisiv 

My suggestion to downgrade was meant to be a workaround to get back _SaveAsAdmin_ functionality as soon as possible in contrast to install a fixed version of the plugin that solely prevents Notepad++ from crashing but has no functionality.

Of course, the best solution is to fix the plugin and rollout a new version. But it seems like @Hsilgos currently has no time for that. Thus, my suggestion could be valuable for some people.

**Comment by Hsilgos** [comment]
Hi, thank you all for your report and Investigations. If problem is only in GetSaveFileNameW function then this feature just won't work. 
I need to consider to process such cases and stop throwing exceptions from DllMain :)
I'll investigate soon how to deal with new mechanism. Until then I'll revert the feature soon and make a new release. Sorry for inconvenience.

**Comment by Hsilgos** [comment]
Ok, I merged necessary change. Unfortunatelly the releasing process seems that has significantly changed and I need some time to update repository to support it

**Comment by Hsilgos** [comment]
Release is [ready](https://github.com/Hsilgos/nppsaveasadmin/releases/tag/1.0.211), tomorrow I'll update the nppPluginList

**Comment by dodmi** [comment]
@Hsilgos 
1st: Thank you for your quick reaction and fixing the crash.
2nd: Is there any chance to see the "Save as" feature again (I know, thinking about where the new file should be saved before starting Notepad++ circumvents the issue, but in reality...)

**Comment by mere-human** [comment]
> I'll investigate soon how to deal with new mechanism. Until then I'll revert the feature soon and make a new release. Sorry for inconvenience.

In the current version, Notepad++ uses [Common Item Dialog](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/bb776913(v=vs.85)) to open or save files.
You can check the code for more details:
[PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp#L786](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/WinControls/OpenSaveFileDialog/CustomFileDialog.cpp#L786)

**Comment by mere-human** [comment]
@Hsilgos you can try hooking the `CoCreateInstance` from `ole32.dll` and check if it was called with the `CLSID_FileSaveDialog` or `CLSID_FileOpenDialog`.

**Comment by Hsilgos** [comment]
> @Hsilgos
> 1st: Thank you for your quick reaction and fixing the crash.
> 2nd: Is there any chance to see the "Save as" feature again (I know, thinking about where the new file should be saved before starting Notepad++ circumvents the issue, but in reality...)

If I won't able to hook into new dialog i'll try to add a new menu command to "Plugins->Save As Admin" to run custom dialog. Not perfect, but usable.

**Comment by Hsilgos** [comment]
> @Hsilgos you can try hooking the `CoCreateInstance` from `ole32.dll` and check if it was called with the `CLSID_FileSaveDialog` or `CLSID_FileOpenDialog`.

Thanks, I'll look into it. Hooking is the easiest part, while adding custom logic might be tricky. I was lucky with GetSaveFileNameW where I could  just add additional flag in hooked function, let's see how it gonna be with this COM.

**Comment by mere-human** [comment]
Maybe, it would be easier to incorporate some part of "Save As Admin" functionality directly into Notepad++ code.

**Comment by Hsilgos** [comment]
> Maybe, it would be easier to incorporate some part of "Save As Admin" functionality directly into Notepad++ code.

Yes, IMHO the most correct way is just to embed this functionality directly into Notepad++. I always felt that hooking WinAPI functions is not the best practice to write plugins :). But Notepad++ has its own mechanism to restart with admin privileges and I'm not sure if they apply the way I do it.
The other way is to send additional events from Notepad++ to plugins to let it manipulate with files/filesystem.

**Comment by allenatdecisiv** [comment]
@mere-human and @Hsilgos,

At the risk of adding more noise, BowPad is an editor which has a "Save as Admin" type of feature. BowPad detects an error during save and offers to relaunch itself with Admin privs (UAC Elevation prompt included), then you can save the file.

BowPad appears to be using Windows "runas" to elevate itself:
https://github.com/stefankueng/BowPad/blob/2519bab7cf898d0dfc1eaa9f371b15404ec8b4ff/src/DocumentManager.cpp#L279

Maybe Notepad++ could do the same (e.g. re-launch itself while retaining all the session metadata and open documents)?

**Comment by mere-human** [comment]
> Maybe Notepad++ could do the same (e.g. re-launch itself while retaining all the session metadata and open documents)?

Notepad++ already does that but not in all cases.
For example, if you open an existing file (say `C:\tmp.txt`) then modify it and save - Notepad++ will ask to relaunch as admin.
But if saving a new file to the protected location, it will only suggest you to save to the documents folder which is not that useful.
I will create a separate issue for that.

**Comment by allenatdecisiv** [comment]
@mere-human,

Very helpful! Thank you!

Ah, I see the use case gap. And I just confirmed BowPad does the same thing for a new unsaved buffer:
"Would you like to save in the Documents folder instead?"

I thought about this some more and I don't think I have ever had to create a new file in a UAC-protected path. I am pretty sure I've always had to edit a file that was in a UAC protected folder, which then required me to have elevated privs (or know that I was going to edit that file _prior_ to launching Notepad++, which I always forgot to do ;-)).

Kudos! I can use Notepad++ as is.

**Comment by vertigo220** [comment]
I tried reinstalling this through N++ but it still didn't work, so I didn't even realize it was updated to fix the problem until someone mentioned it on the N++ issue. I'm guessing it hasn't been updated in the N++ repo yet?

**Comment by Hsilgos** [comment]
@vertigo220 I [updated](https://github.com/notepad-plus-plus/nppPluginList/pull/267) N++ plugin list repo, but I'm not sure how it gonna be updated in N++ itself, probably with next version? In any case you may download last version here and unpack to N++ plugin's folder

**Comment by vertigo220** [comment]
I don't know about automatic updating, but I removed the plugin and went into N++'s Plugins Admin window and reinstalled it that way, and it still didn't work, so I assume it didn't install the newest version even doing it manually that way. And this was just an hour or so ago. Updating it by downloading from here and manually adding to the plugin folder did work.

**Comment by slyfox1186** [comment]
The latest version x64 installer doesn't work. Any attempts to open NPP fails. No window or anything happens when double-clicking on the exe file. The 7zip Portable version does work however. 

Can anyone advise me on what I can do to help the devs further a fix for this software breaking bug?

**Comment by Hsilgos** [comment]
I just now realized that this issue will not let N++ start and it doesn't matter that I updated the plugin. To pick up a new version N++ should start, but it can't...
There are changes to N++ itself are necessary.

**Comment by cyclistefou** [comment]
I confirm the issue is still there with Notepad++ 7.9.3 on Windows 10 x64, using the latest plugin from Plugins Admin as of today - it's still the version 1.0.195 that gets installed. So I followed the link 3 posts above, downloaded v1.0.211 manually, and installed manually by unzipping as "NppSaveAsAdmin" and copying (as admin) into plugins folder. Now N++ starts up correctly.

**Comment by Hsilgos** [comment]
I assume that Notepad++ solved issue with smart installer.
